### PR TITLE
chore: bypass transformer call if UDTransformer is not defined

### PR DESCRIPTION
**Pull request by ayildirim21**
related to: #1612

### CODE REVIEW

File: pkg/sources/forward/data_forward.go [hunk 1]
```diff
@@ -278,35 +278,44 @@ func (df *DataForward) forwardAChunk(ctx context.Context) {
...
+				defer wg.Done()
+				df.concurrentApplyTransformer(ctx, transformerCh)
+			}()
+		}
+		concurrentTransformerProcessingStart := time.Now()
+
+		// send to transformer only the data messages
+		for idx, m := range readMessages {
+			// emit message size metric
+			metrics.ReadBytesCount.With(map[string]string{metrics.LabelVertex: df.vertexName, metrics.LabelPipeline: df.pipelineName, metrics.LabelVertexType: string(dfv1.VertexTypeSource), metrics.LabelVertexReplicaIndex: strconv.Itoa(int(df.vertexReplica)), metrics.LabelPartitionName: df.reader.GetName()}).Add(float64(len(m.Payload)))
```
Comment (KeranYang) [inline pkg/sources/forward/data_forward.go:297]: We need this metric even when transformer is not specified

Comment (KeranYang) [reply pkg/sources/forward/data_forward.go:297]: Also, the watermark assignment right below is needed.

Comment (KeranYang) [reply pkg/sources/forward/data_forward.go:297]: Let me know if you are not sure about which piece is transformer-specific or global.

File: pkg/sources/forward/data_forward.go [hunk 2]
```diff
@@ -192,6 +192,31 @@ type readWriteMessagePair struct {
 	transformerError error
 }
 
+func processReadMessages(hasUDTransformer bool, readMessages []*isb.ReadMessage, processorWM wmb.Watermark,
```
Comment (KeranYang) [inline pkg/sources/forward/data_forward.go:195]: Please write a small paragraph as a comment about what this function does, especially what it **owns**. If it owns processing the messages from the ReadMessages list and generating WriteMessages list ready for writing to buffer, then strictly there should be no concept of UDTransformer outside of this function.

File: pkg/sources/forward/data_forward.go [hunk 3]
```diff
@@ -192,6 +192,31 @@ type readWriteMessagePair struct {
...
+		// assign watermark to the message
+		m.Watermark = time.Time(processorWM)
+		transformerResults[idx].readMessage = m
+
+		if hasUDTransformer {
+			// send transformer processing work to the channel
+			transformerCh <- &transformerResults[idx]
+		} else {
+			// if no user-defined transformer exists, then the messages to write will be identical to the messages read from source
+			transformerResults[idx].writeMessages = []*isb.WriteMessage{{Message: m.Message}}
```
Comment (KeranYang) [inline pkg/sources/forward/data_forward.go:215]: * With this change, is transformerResults still a concise name?
* It looks we are assigning writeMessages within processReadMessages. Does `processReadMessages` own populating writeMessages? If so, should it also populate when a transformer exists?

We can sync up on this tmr morning in the office.

**Comment by KeranYang** [comment]
@ayildirim21, let's also think about how we can unit-test this change.

File: pkg/sources/forward/data_forward.go [hunk 4]
```diff
@@ -270,50 +323,52 @@ func (df *DataForward) forwardAChunk(ctx context.Context) {
...
+
+		metrics.SourceTransformerConcurrentProcessingTime.With(map[string]string{
+			metrics.LabelVertex:             df.vertexName,
+			metrics.LabelPipeline:           df.pipelineName,
+			metrics.LabelVertexReplicaIndex: strconv.Itoa(int(df.vertexReplica)),
+			metrics.LabelPartitionName:      df.reader.GetName(),
+		}).Observe(float64(time.Since(concurrentTransformerProcessingStart).Microseconds()))
+		// transformer processing is done.
+	} else {
+		processReadMessages(false, readMessages, processorWM, readWriteMessagePairs, df, nil)
```
Comment (KeranYang) [inline pkg/sources/forward/data_forward.go:363]: This function call has `hasUDTransformer` as `false` and `tranformerCh` as `nil`. I sense an opportunity for improvement. Can we include line L328->L364 in the processReadMessages method such that we can call `processReadMessages(df.transformer != nil, readMessages, processorWM, readWriteMessagePairs, df)` and inside of `processReadMessages`, conditionally create a transformerCh?

Two benefits: 1. more readable code as developers might not care about the transformer when looking through the code. 2. de-coupled input parameters of `processReadMessages`.

Comment (ayildirim21) [reply pkg/sources/forward/data_forward.go]: I refactored a bit and if we create the channel inside the helper function we would essentially just move the entire logic into the function, so i decided to just keep it in the forwardAChunk method and make a helper for emitting the metrics.

File: pkg/sources/forward/data_forward.go [hunk 5]
```diff
@@ -192,6 +195,17 @@ type readWriteMessagePair struct {
 	transformerError error
 }
 
+// emitMessageSizeMetric emits the message size metric
+func emitMessageSizeMetric(df *DataForward, m *isb.ReadMessage) {
```
Comment (KeranYang) [inline pkg/sources/forward/data_forward.go:199]: I don't think we need to create a separate function for `ReadBytesCount`. For 2 reasons: 1. it's inconsistent with how we emit other metrics, one can argue if we create func for ReadBytesCount, why not all other metrics? 2. and mainly, functions should be [deep](https://medium.com/@alexturdean/shallow-functions-the-silent-killer-of-good-software-527a29ab56c9).

**Review by KeranYang — CHANGES_REQUESTED** [review]

File: pkg/sources/source.go [hunk 6]
```diff
@@ -271,7 +270,7 @@ func (sp *SourceProcessor) Start(ctx context.Context) error {
 	if sp.VertexInstance.Vertex.HasUDTransformer() {
 		sourceForwarder, err = sourceforward.NewDataForward(sp.VertexInstance, sourceReader, writersMap, sp.getTransformerGoWhereDecider(shuffleFuncMap), srcTransformerGRPCClient, fetchWatermark, sourceWmPublisher, toVertexWatermarkStores, idleManager, forwardOpts...)
 	} else {
-		sourceForwarder, err = sourceforward.NewDataForward(sp.VertexInstance, sourceReader, writersMap, sp.getSourceGoWhereDecider(shuffleFuncMap), applier.Terminal, fetchWatermark, sourceWmPublisher, toVertexWatermarkStores, idleManager, forwardOpts...)
+		sourceForwarder, err = sourceforward.NewDataForward(sp.VertexInstance, sourceReader, writersMap, sp.getSourceGoWhereDecider(shuffleFuncMap), nil, fetchWatermark, sourceWmPublisher, toVertexWatermarkStores, idleManager, forwardOpts...)
 	}
```
Comment (yhl25) [inline pkg/sources/source.go:274]: make transformer an optional parameter? you can refer to fallback sink implementation.

File: pkg/sources/forward/data_forward.go (hunk 5 above)
Comment (yhl25) [reply pkg/sources/forward/data_forward.go:199]: +1

**Comment by KeranYang** [comment]
@ayildirim21 awesome job on collecting the perf data!

**Review by KeranYang — APPROVED** [review]

**Review by vigith — APPROVED** [review]

**Comment by vigith** [comment]
@yhl25, can you look into why idle-source for kafka is failing?

**Review by yhl25 — APPROVED** [review]

**Pull request merged**